packages/client/src/sync/fetcher/trienodefetcher.ts

```diff
@@ -0,0 +1,520 @@
+   * All keys in pathToNodeRequestData are sync paths.
+   */
+  pathToNodeRequestData: OrderedMap<string, NodeRequestData>
+
+  // Holds active requests to remove after storing
+  requestedNodeToPath: Map<string, string>
+  fetchedAccountNodes: Map<string, FetchedNodeData> // key is node hash
+  accountTrie: Trie
+  codeTrie: Trie
+  accountToStorageTrie: Map<String, Trie>
+  nodeCount: number
+
+  /**
+   * Create new block fetcher
+   */
+  constructor(options: TrieNodeFetcherOptions) {
+    super(options)
+    this.root = options.root
+    this.pathToNodeRequestData = new OrderedMap<string, NodeRequestData>()
+    this.requestedNodeToPath = new Map<string, string>()
+    this.fetchedAccountNodes = new Map<string, FetchedNodeData>()
+    this.accountTrie = options.accountTrie ?? new Trie({ useKeyHashing: false })
+    this.codeTrie = options.codeTrie ?? new Trie({ useKeyHashing: false })
+    this.accountToStorageTrie = options.accountToStorageTrie ?? new Map<String, Trie>()
+    this.nodeCount = 0
+    this.debug = createDebugLogger('client:TrieNodeFetcher')
+
+    // will always start with root node as first set of node requests
+    this.pathToNodeRequestData.setElement('', {
+      nodeHash: bytesToHex(this.root),
+      nodeParentHash: '', // root node does not have a parent
+    } as NodeRequestData)
+
+    this.debug(
+      `Trie node fetcher instantiated with ${this.pathToNodeRequestData.size()} node requests destroyWhenDone=${
+        this.destroyWhenDone
+      }`
+    )
+  }
+
+  setDestroyWhenDone() {
+    this.destroyWhenDone = true
+  }
+
+  /**
+   * Request results from peer for the given job.
+   * Resolves with the raw result
+   * If `undefined` is returned, re-queue the job.
+   * @param job
+   * @param peer
+   */
+  async request(
+    job: Job<JobTask, Uint8Array[], Uint8Array>
+  ): Promise<TrieNodesResponse | undefined> {
+    const { task, peer } = job
+    const { paths, pathStrings } = task
+
+    const rangeResult = await peer!.snap!.getTrieNodes({
+      root: this.root,
+      paths,
+      bytes: BigInt(this.config.maxRangeBytes),
+    })
+
+    // Response is valid, but check if peer is signalling that it does not have
+    // the requested data. For bytecode range queries that means the peer is not
+    // yet synced.
+    const requestedNodeCount = pathStrings.length
+    if (rangeResult === undefined || requestedNodeCount < rangeResult.nodes.length) {
+      this.debug(`Peer rejected trienode request`)
+
+      return undefined
+    }
+
+    try {
+      // While results are in the same order as requested hashes but there could be gaps/misses in the results
+      // if the node doesn't has the bytecode. We need an index to move forward through the hashes which are
+      // absent in the receieved responses
+      const receivedNodes: Uint8Array[] = []
+      const requestedNodes = new Set(Array.from(this.requestedNodeToPath.keys()))
```

Comment:
`requestedNodes` is a subset of `pathToNodeRequest` and keeps track of all the paths that are actively being requested so that we can cross-reference received nodes to make sure a peer isn't sending us nodes that haven't been requested.

Reply:
fixed